### PR TITLE
memorymanager: fix checkpoint file comparison

### PR DESCRIPTION
**Pull request by Tal-or**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

For a resource within a group, such as memory,
we should validate the total `Free` and total `Reserved` size of the expected `machineState` and state restored from checkpoint file after kubelet start.
If total `Free` and total `Reserved` are equal, the restored state is valid.
    
The old comparison however was done by reflection.
    
There're times when the memory accounting is equals
but the allocations across the NUMA nodes are varies.
    
In such cases we still need to consider the states as equals.

#### Which issue(s) this PR fixes:

Fixes #https://github.com/kubernetes/kubernetes/issues/113130 

#### Special notes for your reviewer:

This PR use as a replacement for https://github.com/kubernetes/kubernetes/pull/114501 which is not active for a long time now and is needed for MemoryManager GA graduation.
see: https://github.com/kubernetes/enhancements/issues/1769#issuecomment-2317059714

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1769-memory-manager#design-overview
```

### CODE REVIEW

**Review by ffromani — COMMENTED** [review]
provisional LGTM (need to do another pass).
The ordering of allocation should not matter indeed, only the totals should matter

**Review by ffromani — COMMENTED** [review]
provisional LGTM

since we cannot guarantee we recreate pods in the very same ordering across restarts, there are multiple legal representations (even including pinning) of the same memory state, thus we should check the aggregate values

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 1]
```diff
@@ -682,8 +681,36 @@ func areMachineStatesEqual(ms1, ms2 state.NUMANodeMap) bool {
...
+				tmpState2.Free += ms2[nodeID].MemoryMap[resourceName].Free
+				tmpState2.Reserved += ms2[nodeID].MemoryMap[resourceName].Reserved
+			}
+
+			if tmpState1.Free != tmpState2.Free {
+				klog.InfoS("Memory states for the NUMA node and resource are different", "node", nodeID, "resource", resourceName, "field", "free", "free1", tmpState1.Free, "free2", tmpState2.Free, "memoryState1", *memoryState1, "memoryState2", *memoryState2)
+				return false
+			}
+			if tmpState1.Reserved != tmpState2.Reserved {
+				klog.InfoS("Memory states for the NUMA node and resource are different", "node", nodeID, "resource", resourceName, "field", "reserved", "reserved1", tmpState1.Reserved, "reserved2", tmpState2.Reserved, "memoryState1", *memoryState1, "memoryState2", *memoryState2)
```
[the remark this answers is not in the document]
Comment (Tal-or) [reply pkg/kubelet/cm/memorymanager/policy_static.go]: Done. Thanks!

**Review by ffromani — COMMENTED** [review]
/lgtm

**Review by mrunalp — APPROVED** [review]

**Pull request merged**